### PR TITLE
Add a class QgsOptional

### PR DESCRIPTION
**Pull request by m-kuhn**
Adds a new class QgsOptional. This adds an "enabled" flag to any other datatype (with default and copy constructors).

This is useful for configuration options which can be en- or disabled but shouldn't lose the information when disabled.

It's very similar to QgsDataDefined but broader in applicability. You can also make a
- `QgsOptional<QgsExpression>` : implemented as POC with python bindings as QgsOptionalExpression
- `QgsOptional<int>`
- `QgsOptional<bool>`
- `QgsOptional<QString>`
- `QgsOptional<QVariant>`

```
> optionalExpression = QgsOptionalExpression(QgsExpression('"field" > 1'));
> optionalExpression.enabled()
True
> optionalExpression.setEnabled(False)
> optionalExpression.data().expression()
"field" > 1
> optionalExpression.setEnabled(True)
> optionalExpression.data().expression()
"field" > 1
```

Basically this avoids having to define a "xEnabled" or "xActivated" flag for every single configuration option.

I'm currently building other features based on this, so proposing this for early feedback.

@nyalldawson @wonder-sk @NathanW2

### CODE REVIEW

**Comment by nyalldawson** [comment]
> I'm currently building other features based on this, so proposing this for early feedback.

Mind sharing a bit of what your future plans are? I'm interested if they touch on the properties framework work I've been plugging away at...

**Comment by m-kuhn** [comment]
Expression based tab/groupbox visibility in feature forms.

**Comment by mhugo** [comment]
+1 for an optional type in QGIS.

Is it planned to be used more widely in the API ? That would be nice (for return values for instance, or basically wherever a pointer is used when an optional is needed). If it is the case then some other things may be considered:
* the contained type is required to be default constructible, and is constructed even with a "disabled" optional. boost::optional or std::optional use placement new for that
* move constructors / operator
* operator bool (syntactic sugar)

**Comment by m-kuhn** [comment]
> Is it planned to be used more widely in the API ?

I hope so

> for return values for instance, or basically wherever a pointer is used when an optional is needed

100% inline with my thoughts.

> the contained type is required to be default constructible

Already the case now

> move constructors / operator

As soon as we require CXX11

> operator bool (syntactic sugar)

Good idea

File: src/core/qgsoptionalexpression.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#define QGSOPTIONALEXPRESSION_H
+
+#include "qgsoptional.h"
+#include "qgsexpression.h"
+
+/**
+ * An expression with an additional enabled flag.
+ *
+ * This can be used for configuration options where an expression can be enabled
+ * or diabled but when disabled it shouldn't lose it's information for the case
```
Comment (nyalldawson) [inline src/core/qgsoptionalexpression.h]: diabled -> disabled

File: src/core/qgsexpression.cpp [hunk 2]
```diff
@@ -3502,6 +3502,14 @@ bool QgsExpression::isValid( const QString &text, const QgsExpressionContext *co
   return !exp.hasParserError();
 }
 
+void QgsExpression::setExpression( const QString& expression )
+{
+  detach();
+  d->mRootNode = ::parseExpression( expression, d->mParserErrorString );
+  d->mEvalErrorString = QString();
+  d->mExp = expression;
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: Would be good to have a test for this

File: src/core/qgsexpression.h [hunk 3]
```diff
@@ -129,8 +129,22 @@ class CORE_EXPORT QgsExpression
      * it does not need to be re-parsed.
      */
     QgsExpression& operator=( const QgsExpression& other );
+
+    /**
+     * Create an empty expression
+     */
+    QgsExpression();
```
Comment (nyalldawson) [inline src/core/qgsexpression.h]: Is there any way to test if an expression is empty?

**Comment by nyalldawson** [comment]
I'd love to see how you are using this to get a better feel for it... are any of your follow up commits public?

File: src/core/qgsexpression.h (hunk 3 above)
Comment (m-kuhn) [inline src/core/qgsexpression.h]: No root node.
But maybe we can do a valid() and operator bool() method.

Comment (nyalldawson) [inline src/core/qgsexpression.h]: Would be a good idea... I can't even remember why but last week I was looking for a way to do this in current API

**Comment by m-kuhn** [comment]
Here's some [work in progress](https://github.com/m-kuhn/QGIS/commit/e800ce5f00056683ff899d0888874ed2c73d9803#diff-9003f720a9ac83de83615753cbbe469aR233)

**Comment by nyalldawson** [comment]
Ok - LGTM, i can definitely see the value here

File: python/core/qgsexpression.sip [hunk 4]
```diff
@@ -12,8 +12,23 @@ class QgsExpression
...
+
     ~QgsExpression();
 
+    /**
+     * Checks if this expression is valid.
+     * A valid expression could be parsed but does not necessarily evaluate properly.
+     *
+     * @note Added in QGIS 3.0
+     */
+    bool valid() const;
```
Comment (nyalldawson) [inline python/core/qgsexpression.sip]: For consistency - could this be isValid() and we shove the static method off to something like "isValidExpressionString"?

**Comment by nyalldawson** [comment]
> Is the test failing because of the template class?

Possibly - I'd first try explicitly adding the \brief tag before the brief and see if that helps

File: python/core/qgsexpression.sip (hunk 4 above)
Comment (m-kuhn) [inline python/core/qgsexpression.sip]: Ok. Do you like `checkExpression`?

Comment (nyalldawson) [inline python/core/qgsexpression.sip]: Sounds good!

**Comment by m-kuhn** [comment]
Superseded by #3472

**Comment by nyalldawson** [comment]
@m-kuhn do you think it's worth having writeXml/readXml methods in QgsOptionalExpression (or QgsXmlUtils)? I think it could avoid the potential misuse of this class if people only save the expression if the expression is enabled.

**Comment by m-kuhn** [comment]
... and have a widget checkable (collapsible)groupbox that can work with it ...
